### PR TITLE
mingw: re-enable endbr opcode, add workaround for llvm `cet.h` issue

### PR DESCRIPTION
**Pull request by vszakats**
- https://github.com/libressl/openbsd/pull/149 fixes to crash
  on startup issue, thus nuking all endbr opcodes is no longer
  necessary. This effectively enables CET in ASM code for
  MinGW x86_64 builds.
  A broader fix landed now via:
  https://github.com/openbsd/src/commit/db9cd57725789270c36763286ca8dc0a7c1b9d10

- llvm `cet.h` (as of v18) assigns `endbr32` instead of `endbr64`
  to `_CET_ENDBR` macro for MinGW x86_64 targets.
  https://github.com/llvm/llvm-project/blob/llvmorg-18.1.1/clang/lib/Headers/cet.h#L15-L35
  Work this around by mapping `endbr32` to `endbr64`.

Prerequisite: https://github.com/libressl/openbsd/pull/149
Fixes: https://github.com/libressl/portable/issues/1015

### CODE REVIEW

**Comment by busterb** [comment]
Tthanks! We're going to need the change in libressl/openbsd#149 reviewed upstream in OpenBSD, which might take a bit. I might disable endbr* still temporarily while this works its way through, since we have a cert expiration update that's time-sensitive.

File: crypto/CMakeLists.txt [hunk 1]
```diff
@@ -193,7 +193,7 @@ if(HOST_ASM_MINGW64_X86_64)
 		whrlpool/wp-mingw64-x86_64.S
 		cpuid-mingw64-x86_64.S
 	)
-	add_definitions(-Dendbr64=)
```
Comment (busterb) [inline crypto/CMakeLists.txt:196]: Don't forget crypto/Makefile.am.mingw64-x86_64 too

Comment (vszakats) [reply crypto/CMakeLists.txt:196]: Thanks for the reminder, I applied the same for autotools.

**Review by busterb — COMMENTED** [review]
This looks fine to me. @botovq and friends, let us know how you want to handle the upstream change

**Comment by botovq** [comment]
The removal of `_CET_ENDBR` will need more thinking. It needs to stay in `x86_64cpuid.pl` for at least OpenBSD.

We probably need to intercept `endbr64` in `x68_64-xlate.pl` and define it away if it appears in the `.init` section for MinGW, maybe around here.

https://github.com/openbsd/src/blob/master/lib/libcrypto/perlasm/x86_64-xlate.pl#L512

The current theory is that something is computing relocations on all opcodes, where it only expects `call`, leading to the crash.

**Comment by vszakats** [comment]
> The removal of `_CET_ENDBR` will need more thinking. It needs to stay in `x86_64cpuid.pl` for at least OpenBSD.
> 
> We probably need to intercept `endbr64` in `x68_64-xlate.pl` and define it away if it appears in the `.init` section for MinGW, maybe around here.

MSVC CRT also expects a list a function pointers in its init section:
https://learn.microsoft.com/cpp/c-runtime-library/crt-initialization

**Comment by botovq** [comment]
We will most likely remove more from this file which will solve your
problem. It will take a little bit of time. Thanks!

**Comment by vszakats** [comment]
This will have to be rebased on the 3.9.1 tree, once it lands in this repo.

**Pull request merged**